dm/worker/upgrade.go

```diff
@@ -0,0 +1,228 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package worker
+
+import (
+	"encoding/json"
+	"os"
+
+	"github.com/pingcap/errors"
+	"github.com/syndtr/goleveldb/leveldb"
+
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/utils"
+)
+
+const (
+	// The current internal version number of DM-worker used when upgrading from an older version, and it's different from the release version.
+	// NOTE: +1 when an incompatible problem is introduced.
+	currentWorkerInternalNo uint64 = 1
+)
+
+var (
+	// The key used when saving the version of DM-worker
+	dmWorkerVersionKey = []byte("!DM-worker!version")
+	// The default previous version of DM-worker if no valid version exists in DB before the upgrade.
+	defaultPreviousWorkerVersion = version{InternalNo: 0, ReleaseVersion: "None"}
+	// all versions exists in the history.
+	workerVersion1 = version{InternalNo: 1, ReleaseVersion: "v1.0.0-alpha"}
+)
+
+// The version of DM-worker used when upgrading from an older version.
+type version struct {
+	InternalNo     uint64 `json:"internal-no"`     // internal version number
+	ReleaseVersion string `json:"release-version"` // release version, like `v1.0.0`
+}
+
+// newVersion creates a new instance of version.
+func newVersion(internalNo uint64, releaseVersion string) version {
+	return version{
+		InternalNo:     internalNo,
+		ReleaseVersion: releaseVersion,
+	}
+}
+
+// newCurrentVersion creates a new instance of version match the current DM-worker.
+func newCurrentVersion() version {
+	return newVersion(currentWorkerInternalNo, utils.ReleaseVersion)
+}
+
+// compare compares the version with another version.
+// NOTE: also compare `ReleaseVersion` when needed.
+func (v *version) compare(other version) int {
+	if v.InternalNo < other.InternalNo {
+		return -1
+	} else if v.InternalNo == other.InternalNo {
+		return 0
+	}
+	return 1
+}
+
+// String implements Stringer.String.
+func (v version) String() string {
+	data, err := v.MarshalBinary()
+	if err != nil {
+		log.Errorf("[worker upgrade] marshal version (internal-no: %d, release-version: %s) to binary error %v",
+			v.InternalNo, v.ReleaseVersion, err)
+		return ""
+	}
+	return string(data)
+}
+
+// MarshalBinary implements encoding.BinaryMarshal.
+func (v *version) MarshalBinary() ([]byte, error) {
+	return json.Marshal(v)
+}
+
+// UnmarshalBinary implements encoding.BinaryMarshal.
+func (v *version) UnmarshalBinary(data []byte) error {
+	return json.Unmarshal(data, v)
+}
+
+// loadVersion loads the version of DM-worker from the levelDB.
+func loadVersion(h Getter) (ver version, err error) {
+	if whetherNil(h) {
+		return ver, errors.Trace(ErrInValidHandler)
+	}
+
+	data, err := h.Get(dmWorkerVersionKey, nil)
+	if err != nil {
+		if err == leveldb.ErrNotFound {
+			return defaultPreviousWorkerVersion, nil
```

Comment:
Is it needed to add a log here?
DM-worker may `saveVersion` failed but leave a valid levelDB directory.

Reply:
added in https://github.com/pingcap/dm/pull/187/commits/10281b801665c53de0201cd2cb9a1a82dc2ccec7.